source/websocket.c

```diff
@@ -876,22 +918,163 @@ static int s_handler_process_read_message(
+    int err;
+
+    websocket->thread_data.incoming_message_window_update = message->message_data.len;
+
+    AWS_LOGF_TRACE(
+        AWS_LS_HTTP_WEBSOCKET,
+        "id=%p: Begin processing message of size %zu.",
+        (void *)websocket,
+        message->message_data.len);
+
+    while (cursor.len) {
+        if (websocket->thread_data.is_reading_stopped) {
+            goto clean_up;
+        }
+
+        bool frame_complete;
+        err = aws_websocket_decoder_process(&websocket->thread_data.decoder, &cursor, &frame_complete);
+        if (err) {
+            AWS_LOGF_ERROR(
+                AWS_LS_HTTP_WEBSOCKET,
+                "id=%p: Message processing failed, error %d (%s). Closing connection.",
+                (void *)websocket,
+                aws_last_error(),
+                aws_error_name(aws_last_error()));
+
+            goto error;
+        }
+
+        if (frame_complete) {
+            s_complete_incoming_frame(websocket, AWS_ERROR_SUCCESS);
+        }
+    }
+
+    if (websocket->thread_data.incoming_message_window_update > 0) {
+        err = aws_channel_slot_increment_read_window(slot, websocket->thread_data.incoming_message_window_update);
+        if (err) {
+            AWS_LOGF_ERROR(
+                AWS_LS_HTTP_WEBSOCKET,
+                "id=%p: Failed to increment read window after message processing, error %d (%s). Closing connection.",
+                (void *)websocket,
+                aws_last_error(),
+                aws_error_name(aws_last_error()));
+            goto error;
+        }
+    }
+
+    goto clean_up;
+
+error:
+    s_shutdown_due_to_read_err(websocket, aws_last_error());
+
+clean_up:
+    if (cursor.len > 0) {
+        AWS_LOGF_TRACE(
+            AWS_LS_HTTP_WEBSOCKET,
+            "id=%p: Done processing message, final %zu bytes ignored.",
+            (void *)websocket,
+            cursor.len);
+    } else {
+        AWS_LOGF_TRACE(AWS_LS_HTTP_WEBSOCKET, "id=%p: Done processing message.", (void *)websocket);
+    }
+    aws_mem_release(message->allocator, message);
+    return AWS_OP_SUCCESS;
 }
 
 static int s_decoder_on_frame(const struct aws_websocket_frame *frame, void *user_data) {
-    (void)frame;
-    (void)user_data;
-    return aws_raise_error(AWS_ERROR_UNIMPLEMENTED);
+    struct aws_websocket *websocket = user_data;
+    assert(aws_channel_thread_is_callers_thread(websocket->channel_slot->channel));
+    assert(!websocket->thread_data.has_incoming_frame);
+
+    websocket->thread_data.has_incoming_frame = true;
+
+    struct aws_websocket_incoming_frame *incoming_frame = &websocket->thread_data.incoming_frame;
+    incoming_frame->payload_length = frame->payload_length;
+    incoming_frame->opcode = frame->opcode;
+    incoming_frame->fin = frame->fin;
+    incoming_frame->rsv[0] = frame->rsv[0];
+    incoming_frame->rsv[1] = frame->rsv[1];
+    incoming_frame->rsv[2] = frame->rsv[2];
+
+    /* Invoke user cb */
+    if (websocket->on_incoming_frame_begin) {
```

Comment:
If reading is stopped, should we even invoke this callback?

Reply:
If reading was stopped, we shouldn't get to this part of the code (wouldn't run decoder on incoming bytes, wouldn't get callback about frame-header). I added an assert to make sure that contract is being followed.

Checking is_reading_stopped is in case the user tries to close the websocket from the callback, we want that to stop the decoder from running over any further bytes.